### ISSUE TITLE
Google Cloud Storage: Can't download files with Content-Encoding: gzip

### ISSUE THREAD

**Issue by kevinoid** [rclone/rclone#2658]
#### What is the problem you are having with rclone?

rclone is unable to download a file from Google Cloud Storage which has `Content-Encoding: gzip` due to size mismatch.  (Or MD5 mismatch when copying to Azure.)

#### What is your rclone version (output from `rclone version`)

Reproduced with both the Debian-packaged version:

    rclone v1.41
    - os/arch: linux/amd64
    - go version: go1.10.1

and built from current git `master`:

    rclone v1.44-001-g67703a73-beta
    - os/arch: linux/amd64
    - go version: go1.10.4

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Debian GNU/Linux 4.18.0 x86_64

####  Which cloud storage system are you using? (eg Google Drive)

Google Cloud Storage

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

    echo 'Example content.' > file.txt
    gsutil cp -Z file.txt gs://$bucket/file.txt.gz
    rclone -vv copy $gcs_remote:$bucket/file.txt.gz file.txt.gz

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

    2018/10/15 15:46:26 DEBUG : rclone: Version "v1.44-001-g67703a73-beta" starting with parameters ["./rclone" "-vv" "copy" "gcs:9718a7ca-c0d4-41ac-a0dc-46922b9d541d/file.txt.gz" "file.txt.gz"]
    2018/10/15 15:46:26 DEBUG : Using config file from "/home/kevin/.config/rclone/rclone.conf"
    2018/10/15 15:46:27 DEBUG : file.txt.gz: Couldn't find file - need to transfer
    2018/10/15 15:46:27 ERROR : file.txt.gz: corrupted on transfer: sizes differ 47 vs 17
    2018/10/15 15:46:27 INFO  : file.txt.gz: Removing failed copy
    2018/10/15 15:46:27 ERROR : Attempt 1/3 failed with 1 errors and: corrupted on transfer: sizes differ 47 vs 17
    2018/10/15 15:46:27 DEBUG : file.txt.gz: Couldn't find file - need to transfer
    2018/10/15 15:46:28 ERROR : file.txt.gz: corrupted on transfer: sizes differ 47 vs 17
    2018/10/15 15:46:28 INFO  : file.txt.gz: Removing failed copy
    2018/10/15 15:46:28 ERROR : Attempt 2/3 failed with 1 errors and: corrupted on transfer: sizes differ 47 vs 17
    2018/10/15 15:46:28 DEBUG : file.txt.gz: Couldn't find file - need to transfer
    2018/10/15 15:46:28 ERROR : file.txt.gz: corrupted on transfer: sizes differ 47 vs 17
    2018/10/15 15:46:28 INFO  : file.txt.gz: Removing failed copy
    2018/10/15 15:46:28 ERROR : Attempt 3/3 failed with 1 errors and: corrupted on transfer: sizes differ 47 vs 17
    2018/10/15 15:46:28 Failed to copy: corrupted on transfer: sizes differ 47 vs 17

I'm guessing that the problem is that GCS reports `Content-Length: 47` (the length of the gzip-encoded content) while rclone is using 17 decompressed bytes.  (Note:  The compressed content is actually larger due to format overhead.)  Perhaps a call to [`ReadCompressed(true)`](https://godoc.org/cloud.google.com/go/storage#ObjectHandle.ReadCompressed) to disable decompression by the Google Client Library would be appropriate?

Thanks,
Kevin

**Comment by ncw** [comment]
Try this flag `--no-gzip-encoding` not sure it will work, but fixing the above was its original purpose.

**Comment by kevinoid** [comment]
Thanks for the suggestion.  It didn't work for me.  Updated log below:

    2018/10/16 10:08:59 DEBUG : rclone: Version "v1.44-001-g67703a73-beta" starting with parameters ["./rclone" "-vv" "--no-gzip-encoding" "copy" "kevinish:9718a7ca-c0d4-41ac-a0dc-46922b9d541d/file.txt.gz" "file.txt.gz"]
    2018/10/16 10:08:59 DEBUG : Using config file from "/home/kevin/.config/rclone/rclone.conf"
    2018/10/16 10:09:00 DEBUG : file.txt.gz: Couldn't find file - need to transfer
    2018/10/16 10:09:00 ERROR : file.txt.gz: corrupted on transfer: sizes differ 47 vs 17
    2018/10/16 10:09:00 INFO  : file.txt.gz: Removing failed copy
    2018/10/16 10:09:00 ERROR : Attempt 1/3 failed with 1 errors and: corrupted on transfer: sizes differ 47 vs 17
    2018/10/16 10:09:00 DEBUG : file.txt.gz: Couldn't find file - need to transfer
    2018/10/16 10:09:00 ERROR : file.txt.gz: corrupted on transfer: sizes differ 47 vs 17
    2018/10/16 10:09:00 INFO  : file.txt.gz: Removing failed copy
    2018/10/16 10:09:00 ERROR : Attempt 2/3 failed with 1 errors and: corrupted on transfer: sizes differ 47 vs 17
    2018/10/16 10:09:00 DEBUG : file.txt.gz: Couldn't find file - need to transfer
    2018/10/16 10:09:00 ERROR : file.txt.gz: corrupted on transfer: sizes differ 47 vs 17
    2018/10/16 10:09:00 INFO  : file.txt.gz: Removing failed copy
    2018/10/16 10:09:00 ERROR : Attempt 3/3 failed with 1 errors and: corrupted on transfer: sizes differ 47 vs 17
    2018/10/16 10:09:00 Failed to copy: corrupted on transfer: sizes differ 47 vs 17

For reference, I think that flag would be counter-productive since the downloaded content needs to be gzipped so the MD5 and size can be verified.

**Comment by ncw** [comment]
Here it is obvious that the file has been decompressed

```
2018/10/16 18:55:11 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2018/10/16 18:55:11 DEBUG : file2.txt.gz: Couldn't find file - need to transfer
2018/10/16 18:55:11 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2018/10/16 18:55:11 DEBUG : HTTP REQUEST (req 0xc0004e8200)
2018/10/16 18:55:11 DEBUG : GET /download/storage/v1/b/rclone-test1/o/file2.txt.gz?generation=1539712477065527&alt=media HTTP/1.1
Host: www.googleapis.com
User-Agent: rclone/v1.44-DEV
Authorization: XXXX
Accept-Encoding: gzip

2018/10/16 18:55:11 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2018/10/16 18:55:12 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2018/10/16 18:55:12 DEBUG : HTTP RESPONSE (req 0xc0004e8200)
2018/10/16 18:55:12 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Alt-Svc: quic=":443"; ma=2592000; v="44,43,39,35"
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Disposition: attachment
Content-Language: en
Content-Type: text/plain
Date: Tue, 16 Oct 2018 17:55:11 GMT
Etag: CLfigfrDi94CEAE=
Expires: Mon, 01 Jan 1990 00:00:00 GMT
Pragma: no-cache
Server: UploadServer
Vary: Origin
Vary: X-Origin
X-Goog-Generation: 1539712477065527
X-Goog-Hash: crc32c=Aa5TQQ==,md5=551nL1Dz9AYIgwOywlL6iQ==
X-Goog-Metageneration: 1
X-Goog-Storage-Class: STANDARD
X-Guploader-Uploadid: AEnB2Uqsv4GHCnD6G9bokmvuP4OgReLq9Sz-lBSAoKNDMM7Tobxo_QgQlj_J8MkCpIwdlvt7wxBIhEtPBFIpkqEh4JQZyGQYxw

11
Example content.

0

2018/10/16 18:55:12 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2018/10/16 18:55:12 ERROR : file2.txt.gz.download: corrupted on transfer: sizes differ 47 vs 17
```

When I compress the file with gzip first and upload it, it download fine with rclone.

I think the difference is in the Content-Types

```
$ rclone lsf --format mp gcs:rclone-test1
application/gzip;file.txt.gz  # uploaded with rclone
text/plain;file2.txt.gz # uploaded with gsutil -X
```

So it appears that the rclone auto decompressed the file (confusingly from 47 to 17 bytes), but as you said the md5sum of the file is for the compressed data, not the uncompressed.

Here is the help from gsutil which is enlightening

```
  -z <ext,...>   Applies gzip content-encoding to file uploads with the given
                 extensions. This is useful when uploading files with
                 compressible content (such as .js, .css, or .html files)
                 because it saves network bandwidth and space in Google Cloud
                 Storage, which in turn reduces storage costs.

                 When you specify the -z option, the data from your files is
                 compressed before it is uploaded, but your actual files are
                 left uncompressed on the local disk. The uploaded objects
                 retain the Content-Type and name of the original files but are
                 given a Content-Encoding header with the value "gzip" to
                 indicate that the object data stored are compressed on the
                 Google Cloud Storage servers.

                 For example, the following command:

                   gsutil cp -z html -a public-read cattypes.html gs://mycats

                 will do all of the following:

                 - Upload as the object gs://mycats/cattypes.html (cp command)
                 - Set the Content-Type to text/html (based on file extension)
                 - Compress the data in the file cattypes.html (-z option)
                 - Set the Content-Encoding to gzip (-z option)
                 - Set the ACL to public-read (-a option)
                 - If a user tries to view cattypes.html in a browser, the
                   browser will know to uncompress the data based on the
                   Content-Encoding header, and to render it as HTML based on
                   the Content-Type header.

                 Note that if you download an object with Content-Encoding:gzip
                 gsutil will decompress the content before writing the local
                 file.
```

So I think `--no-gzip-encoding` should be doing the right thing... Unfortunately it doesn't seem to be working as I can see rclone is still outputing `Accept-Encoding: gzip` - perhaps the SDK is overriding that.

However `--no-gzip-encoding` would download the file compressed which probably isn't what the user expects.

What would you expect rclone to do? Decompress the file?

**Comment by kevinoid** [comment]
> What would you expect rclone to do? Decompress the file?

My original goal was to transfer files from Google Cloud Storage to Azure Storage, so ideally it would transfer the compressed content exactly and preserve the Content-Encoding and MD5 metadata in the destination.

When copying a file from GCS to the local filesystem (or any where Content-Encoding is not declared), I think it makes the most sense to save the compressed content, both for performance reasons and so the MD5 can be verified (and preserved end-to-end in multi-step copies).  However, I could see a case for automatic decompression upon saving since `Content-Encoding` is often transparent to users (such as the use case described in the `gsutil` help you quoted).

Having a flag for user-selectable behavior might not be a bad idea.  I'm not sure which is the more reasonable default.  Perhaps the "automatic" default could choose based on the presence of a `.gz` extension?

**Comment by ncw** [comment]
I found a great article explaining this on gcs: https://cloud.google.com/storage/docs/transcoding

So what I need to do is add the "Accept-Encoding: gzip" header which will have two effects
- gcs serves the file compressed
- the go transport won't automatically decompress it

I only add that for files with Content-Encoding: gzip which means I think it is safe to leave on permanently.

It seems likely that s3 has the same problem, but I haven't found their docs page yet!

I've made a beta with that in  here:

https://beta.rclone.org/branch/v1.44-003-g748b2fed-fix-2658-gcs-gzip-beta/ (uploaded in 15-30 mins)

**Comment by kevinoid** [comment]
That sounds great!  I tested the beta and it successfully downloaded the compressed file and preserved the MD5.  Excellent!  Thank you!

However, I noticed that it doesn't preserve the Content-Encoding (or custom metadata) when copying from GCS to Azure.  Should I open a separate issue for that, or would you like to discuss in this one?  In either case, I will write up more details and a test case.

Thanks again!

**Comment by ncw** [comment]
> That sounds great! I tested the beta and it successfully downloaded the compressed file and preserved the MD5. Excellent! Thank you!

:-)
 
> However, I noticed that it doesn't preserve the Content-Encoding (or custom metadata) when copying from GCS to Azure. Should I open a separate issue for that, or would you like to discuss in this one? 

There are two sorts of metadata, general purpose key, value storage and what I'll call HTTP metadata
- Access control metadata
- Cache-Control
- Content-Disposition
- Content-Encoding
- Content-Language
- Content-Type

Rclone deals with `Content-Type` already but it doesn't deal with the other kinds of metadata.

There is an issue already about custom metadata #111.  That is quite an old issue but I think it would be much easier to implement now-a-days.  You'll see various other issues linked from there.

It would be nice if rclone could copy metadata from cloud to cloud, and also set it on upload.  This would be a reasonably big project though!

The way it would work is that I'd give each Object an optional interface `ReadMetadata` and that would be supplied.  On upload this would be read and set on the object.

> In either case, I will write up more details and a test case.

That would be useful - can you put it in #111.  I'll move that issue up into the run Q since I think its time has come :-)

I think the only thing I'm not sure about is how to represent the http metadata and the non http metadata in a cross cloud sort of way.  Perhaps ReadMetadata should return two dictionaries or one dictionary and one `http.Header`

Thinking aloud this could also subsume the `Content-Type` mechanism.

Docs
- gcs: https://cloud.google.com/storage/docs/metadata
- azureblob: https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob-properties
- s3: https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html

This could also (with a flag) be implemented as attributes on local files.

**Comment by kevinoid** [comment]
That sounds great.  Thanks for the detailed explanation!

**Comment by kgriffs** [comment]
It seems this may have broken again, since I just ran into it myself. I have some JSON files that were stored compressed in Google Storage with `Content-Encoding` set to `gzip` in the metadata.

When running `rclone sync -P` from the bucket to a local directory, I get the `corrupted on transfer: sizes differ` error for each of these files, presumably because rclone is checking the size of the object in the bucket against the decompressed size after download.

Passing `--header-download "Accept-Encoding: gzip"` works around the error, but now I end up with a bunch of `.json` files that are actually still gzip'ed, which is not what I would expect. Rather, I would like them to be stored decompressed. I understand that there are good arguments for storing compressed or not, and so it may be best to provide a flag to toggle this option.

In summary, it would be great if rclone could be updated to support storing the file either compressed or not (depending on a flag), but whatever the value of that option, it performs the size/hash comparison before decompression.

**Comment by ncw** [comment]
> Passing --header-download "Accept-Encoding: gzip" works around the error

That is a good workaround and is equivalent to the patch above.

I had another idea about this here

[v1.56.0-beta.5387.42a7efc4a.fix-2658-gcs-gzip-unknown-size](https://beta.rclone.org/branch/fix-2658-gcs-gzip-unknown-size/v1.56.0-beta.5387.42a7efc4a.fix-2658-gcs-gzip-unknown-size/) on branch [fix-2658-gcs-gzip-unknown-size](https://github.com/rclone/rclone/tree/fix-2658-gcs-gzip-unknown-size) (uploaded in 15-30 mins)

This modifies the patch above and if a gzipped object is detected then
- clear the hash
- mark the object as unknown size

This would let them be downloaded decompressed.

So there are two approaches

1. download the files uncompressed - md5sum not checked, but crc in gzip is checked
2. download the files compressed - md5sum is checked

Do you think rclone should engage 1) automatically with 2) being an option?

**Comment by panthony** [comment]
Encountered the same issue, using `--header-download "Accept-Encoding: gzip"` does not work for me as it fails with:

>  Failed to copy: failed to open source object: SignatureDoesNotMatch: The request signature we calculated does not match the signature you provided. Check your Google secret key and signing method.

I'm syncing from a Cloud Storage using S3 API (Provider: Other, endpoint: https://storage.googleapis.com).

I would have love to be able to retrieve the file compressed (either because uploaded compressed or with Content-Encoding) and copy it elsewhere compressed.

For now the only thing working for me is `--no-gzip-encoding` but I'm now copying the data uncompressed.

**Comment by ncw** [comment]
@panthony I developed the above patch a bit more...

>     gcs: Fix download of "Content-Encoding: gzip" compressed objects
>     
>     Before this change, if an object compressed with "Content-Encoding:
>     gzip" was downloaded, a length and hash mismatch would occur since the
>     as the go runtime automatically decompressed the object on download.
>     
>     This change erases the length and hash on compressed objects so they
>     can be downloaded successfully, at the cost of not being able to check
>     the length or the hash of the downloaded object.
>     
>     This also adds the --gcs-download-compressed flag to allow the
>     compressed files to be downloaded as-is providing compressed objects
>     with intact size and hash information.

So what you'd want is `--gcs-download-compressed` to download the compressed objects only. If you don't use that flag then you'll get uncompressed objects which we can't check the size or hash of.

I think this should work with GCS. You are talking about S3 too - do you need it there too?

**Comment by ncw** [comment]
Here is the beta to try

[v1.59.0-beta.6046.957016195.fix-2658-gcs-gzip-unknown-size](https://beta.rclone.org/branch/fix-2658-gcs-gzip-unknown-size/v1.59.0-beta.6046.957016195.fix-2658-gcs-gzip-unknown-size/) on branch [fix-2658-gcs-gzip-unknown-size](https://github.com/rclone/rclone/tree/fix-2658-gcs-gzip-unknown-size) (uploaded in 15-30 mins)

**Comment by panthony** [comment]
@ncw At the moment:

- I'm always reading using S3 backend (either AWS or GCS @ https://cloud.google.com/storage/docs/interoperability)
- I'm always writing to GCS

I'm actually never reading from GCS directly (not by choice).

And ideally, I'd like the compressed version whenever possible 👍🏻 

Thanks for your suggested patch, it's a good place to start

**Comment by ncw** [comment]
@panthony so you'd really want this patch for the s3 backend is that right?

Are you trying to sync compressed blobs from s3 -> gcs? That might need a bit more magic to preserve the content encoding?

**Comment by panthony** [comment]
> so you'd really want this patch for the s3 backend is that right?

Sure!

> Are you trying to sync compressed blobs from s3 -> gcs? That might need a bit more magic to preserve the content encoding?

I'm trying to sync files (copy really...) from various sources I have no control over into Cloud Storage, thus the files can be:

- uncompressed
- file manually compressed then uploaded
- uncompressed and uploaded with "content-encoding"

Ideally I'd like to copy the compressed version whenever possible (saves both money & time) but a requirement I have is that if I re-run copy on those files they should be detected as "up to date" and not be overwritten.

I do not really care about the Content-Encoding as far as I can tell, those files are handled by backend processing and not serve via a public URL or something alike.


To be noted that I'm currently exclusively relying on checksum (--checksum --ignore-size) because some of the sources I'm copying from have uploaded blobs using chunked encoding and they miss the Content-Length attribute, making rclone fail its check on the size after the upload (size differs 0 vs X).

**Comment by panthony** [comment]
@ncw Well, I have a new occurence, this time from GCS, where I cannot download a file with or without `--no-gzip-encoding`.

In both case it fails the size and/or checksum check.

Your branch `gzip-unknown-size` resolved this issue **AND** I'm able to copy the compressed version which is a huge win so thank you for that.

I wonder if that flag could make its way into the main repository to avoid relying on a forked version of rclone.

**Comment by ncw** [comment]
@panthony I should really merge this, you are right!

I rebased that branch - can you give it a test and if it works OK for you I'll merge it.

[v1.59.0-beta.6166.6bc8b135e.fix-2658-gcs-gzip-unknown-size](https://beta.rclone.org/branch/fix-2658-gcs-gzip-unknown-size/v1.59.0-beta.6166.6bc8b135e.fix-2658-gcs-gzip-unknown-size/) on branch [fix-2658-gcs-gzip-unknown-size](https://github.com/rclone/rclone/tree/fix-2658-gcs-gzip-unknown-size) (uploaded in 15-30 mins)

The same thing needs to be done for s3 too I think.

**Comment by panthony** [comment]
@ncw I can confirm:

- latest version (1.58.1) fail with or without --no-gzip-encoding
- using your branch:
  - download uncompressed version by default successfully
  - detect change in size & download compressed if I re-run using `--gcs-download-compressed`
  - detect that file is identical on the next run

A similar fix would be nice for S3 👍🏻

**Comment by panthony** [comment]
There might be an unexpected side effect though, I though that the file size would be drop **before** the comparison but we do loose the size in listing too, the command:

```
rclone lsl source:bucket
```

Will return `-1` as file size.

**Comment by ncw** [comment]
> There might be an unexpected (for me) side effect though.
> 
> I thought that the file size would be dropped **before** the comparison because we can't rely on it when copying to check if successful but we do loose the size in listing too, the command:
> 
> ```
> rclone lsl source:bucket
> ```
> 
> Will return `-1` as file size, I expected the listing to continue reporting the file size as returned by the backend.

In rclone objects either have a known size, >=0 or an unknown size -1. So rclone doesn't know the size of the Gzip encoded files unfortunately as it can't know how big they really are until they've been decompressed.

If you want to see the compressed size, you'd have to use --gcs-download-compressed

Do you think this is a show stopper?

**Comment by panthony** [comment]
> If you want to see the compressed size, you'd have to use --gcs-download-compressed
> Do you think this is a show stopper?

Awesome! I thought the flag only impact the copy part, not the listing part, I'm perfectly fine with it!

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.59

@panthony if you'd like to see this change for s3 &or azureblob then please make a new issue - thank you.
